### PR TITLE
CI Testing

### PR DESCRIPTION
**Pull request by jborean93**
Added Appveyor, Travis-CI and Coveralls to the requests-ntlm testing suite. Also added a functional test to test out the code with a real IIS server.

Before this can be merged it will require the URL's and tokens used by the 3 services to be updated with the actual requests/requests-ntlm account.

Thanks

Jordan

### CODE REVIEW

**Comment by jborean93** [comment]
@Lukasa 

Here is the first part of the changes of my pull request from yesterday. They are all situated in the testing side and no code in the requests_ntlm package was touched.

Thanks

Jordan

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ env/
 build/
 dist/
 *.egg-info/
+.idea/
```
Comment (Lukasa) [inline .gitignore]: Generally speaking we'd like to keep IDE-specific files out of the global .gitignore. It should probably be added to your global gitignore file.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+language: python
+
+python:
+  - "2.6"
```
Comment (Lukasa) [inline .travis.yml]: We don't support Python 2.6, we can remove it from our testing matrix.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  - "2.7"
+  - "3.3"
+  - "3.4"
+  - "3.5"
+
+# command to install dependencies
+install:
+  - python setup.py install
```
Comment (Lukasa) [inline .travis.yml]: Don't bother running setup.py directly, do `pip install .` instead.

File: README.rst [hunk 4]
```diff
@@ -1,6 +1,15 @@
...
 =============
 
+.. image:: https://travis-ci.org/jborean93/requests-ntlm.svg?branch=master
+    :target: https://travis-ci.org/jborean93/requests-ntlm
+
+.. image:: https://ci.appveyor.com/api/projects/status/nj172p4nkyalwh6r/branch/master?svg=true
+    :target: https://ci.appveyor.com/project/jborean93/requests-ntlm
+
+.. image:: https://coveralls.io/repos/github/jborean93/requests-ntlm/badge.svg?branch=master
+    :target: https://coveralls.io/github/jborean93/requests-ntlm?branch=master
```
Comment (Lukasa) [inline README.rst]: These links probably should point to the organisation URL, rather than your personal one.

Comment (jborean93) [inline README.rst]: Yep, I can easily change the travis-ci and coveralls but would need the token value for appveyor that is set up for the organisation (requests). Not sure who would have permission to give access for that.

File: .travis.yml (hunk 2 above)
Comment (jborean93) [inline .travis.yml]: The setup.py file says it support 2.6 under classifiers unless that isn't what the section is for?

File: tests/functional/test_functional.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+def send_request(url, username, password):
+    """
+    Sends a request to the url with the credentials specified. Returns the final response
+    """
+    session = requests.Session()
+    session.verify = False
+    session.auth = requests_ntlm.HttpNtlmAuth(username, password)
+    request = requests.Request('GET', url)
+    prepared_request = session.prepare_request(request)
+    response = session.send(prepared_request)
```
Comment (Lukasa) [inline tests/functional/test_functional.py]: Any reason we can't just do this using `session.get()`?

File: tests/functional/test_functional.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    return response
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (Lukasa) [inline tests/functional/test_functional.py]: Given that we've written these tests in the py.test style, there's no point having a unittest fallback.

File: tests/functional/test_functional.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+username = '.\\User'
+password = 'Password01'
+http_with_cbt = 'http://127.0.0.1:81/contents.txt'
+http_without_cbt = 'http://127.0.0.1:82/contents.txt'
+https_with_cbt = 'https://127.0.0.1:441/contents.txt'
+https_without_cbt = 'https://127.0.0.1:442/contents.txt'
+expected = 'contents'
+
+class Test_Functional(unittest.TestCase):
```
Comment (Lukasa) [inline tests/functional/test_functional.py]: Don't bother inheriting from `unittest.TestCase`.

File: tests/unit/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+__author__ = 'requests_ntlm'
```
Comment (Lukasa) [inline tests/unit/__init__.py]: We don't need anything in this file.

File: tests/unit/test_requests_ntlm.py [hunk 9]
```diff
@@ -1,14 +1,15 @@
 import unittest
 import requests
 import requests_ntlm
-import test_server
+
+from tests.test_utils import *
```
Comment (Lukasa) [inline tests/unit/test_requests_ntlm.py]: Can we try to avoid the `import *` and import the names specifically?

File: tests/test_server.py [hunk 10]
```diff
@@ -1,15 +1,16 @@
 from flask import Flask,request
 from ntlm3 import ntlm
+from tests.test_utils import *
```
Comment (Lukasa) [inline tests/test_server.py]: Can we avoid `import *` here, and use the name directly instead?

File: tests/unit/test_requests_ntlm.py [hunk 11]
```diff
@@ -38,5 +39,44 @@ def test_new_requests_are_used(self):
...
+        actual_domain = context.domain
+        actual_user = context.username
+
+        assert actual_domain == expected_domain
+        assert actual_user == expected_user
+
 if __name__ == '__main__':
-    unittest.main()
\ No newline at end of file
+    unittest.main()
```
Comment (Lukasa) [inline tests/unit/test_requests_ntlm.py]: We don't need this `unittest` fallback either.

**Comment by Lukasa** [comment]
Thanks for this @jborean93! This is a great start, but I've left some notes inline that might be worth addressing.

File: .travis.yml (hunk 2 above)
Comment (Lukasa) [inline .travis.yml]: Hrm. Let's keep this for now then.

File: README.rst (hunk 4 above)
Comment (Lukasa) [inline README.rst]: Probably me, but I don't have time to do it tonight. Will do it later tonight. =)

File: tests/functional/test_functional.py (hunk 7 above)
Comment (jborean93) [inline tests/functional/test_functional.py]: Sorry I am fairly new to python and testing but I don't understand your point here. If I remote unittest.TestCase the tests will not run. Do you mean replace it with something else?

Comment (Lukasa) [inline tests/functional/test_functional.py]: It'll run fine. You're running the tests via py.test, which doesn't require that the test classes inherit from `unittest.TestCase`.

**Comment by jborean93** [comment]
Have incorporated all the changes mentioned above in the latest commit. The only thing I couldn't do was change the appveyor token as @Lukasa will generate this sometime later.

File: README.rst (hunk 4 above)
Comment (jborean93) [inline README.rst]: Hey @Lukasa have you had time to look at this so we can get this pull request merged?

**Comment by Lukasa** [comment]
You've still left the `.idea` in the `.gitignore`. ;)

**Comment by Lukasa** [comment]
Hrm, creating a proper appveyor account for a GitHub organisation turns out to be moderately frustrating. And right now I am unable to create one for this repo with my personal account.

**Comment by jborean93** [comment]
I just checked .idea isn't in there anymore, pulled down a fresh repo and it is gone. If you wish to skip appveyor I can remove the badge and we can look at it another time but it would be nice to have a real test against IIS to prove NTLM auth works properly.

Looking online very quickly, it is saying for Appveyor to work with organisations you need to sign up with an email and password and not through the Github SSO. From there you can assign permissions to individual users and manage it that way.

**Comment by Lukasa** [comment]
@jborean93 Yeah, but that's not entirely true. =) For example, appveyor tests several repositories that are under the python-hyper organisation without having needed to jump through that hoop.

**Comment by jborean93** [comment]
Let me know if there is anything I can do to help. I am hoping to merge in the changes for the NTLMv2 work and talk about changing the dependency sometime soon.

**Comment by jborean93** [comment]
@Lukasa any updates on this? I would like to hopefully merge in the other changes after this pull request is done soon.

**Comment by jborean93** [comment]
Hey @Lukasa I have removed the appveyor code and have kept just the Travis-CI stuff. Let me know if there was anything I missed or you would like to change.

**Comment by Lukasa** [comment]
Alright, good stuff @jborean93! I'll turn this on today and see how it goes.

**Pull request merged**